### PR TITLE
plan: new plan supports merge join.

### PR DESCRIPTION
**Pull request by hanfei1991**
It's a little different from the old one:
we don't consider join with multiple key to make code simpler.

@shenli @coocood @ilovesoup PTAL

### CODE REVIEW

File: plan/new_physical_plan_builder.go [hunk 1]
```diff
@@ -110,21 +110,60 @@ func (p *LogicalJoin) convert2NewPhysicalPlan(prop *requiredProp) (taskProfile,
...
-	task = prop.enforceProperty(task, p.ctx, p.allocator)
 	return task, p.storeTaskProfile(prop, task)
 }
 
-func (p *LogicalJoin) convert2SemiJoin() (taskProfile, error) {
+func (p *LogicalJoin) preferUseMergeJoin() bool {
+	return p.preferMergeJoin && len(p.EqualConditions) == 1
+}
+
+func (p *LogicalJoin) convert2MergeJoin(prop *requiredProp) (taskProfile, error) {
```
Comment (coocood) [inline plan/new_physical_plan_builder.go]: This function has the assumption that the length of EqualCondtion is one, which creates an unnecessary dependency on the implementation of `preferUseMergeJoin`.

How about rename it to `tryToConvert2MergeJoin`, check preferMergeJoin within this function?

If we can not convert it to merge join, fall back to hash join.

Comment (ilovesoup) [inline plan/new_physical_plan_builder.go]: Actually I don't quite understand why only one condition is used. For t1.a= t2.a and t1.b = t2.b, shouldn't both taken into consideration?

Comment (ilovesoup) [inline plan/new_physical_plan_builder.go]: If for simplicity, please leave a TODO or a record somewhere else for future tracking.

File: plan/new_physical_plan_builder.go [hunk 2]
```diff
@@ -110,21 +110,60 @@ func (p *LogicalJoin) convert2NewPhysicalPlan(prop *requiredProp) (taskProfile,
...
+	rChild := p.children[1].(LogicalPlan)
+	mergeJoin := PhysicalMergeJoin{
+		JoinType:        p.JoinType,
+		EqualConditions: p.EqualConditions,
+		LeftConditions:  p.LeftConditions,
+		RightConditions: p.RightConditions,
+		OtherConditions: p.OtherConditions,
+	}.init(p.allocator, p.ctx)
+	mergeJoin.SetSchema(p.schema)
+	lJoinKey := p.EqualConditions[0].GetArgs()[0].(*expression.Column)
```
Comment (ilovesoup) [inline plan/new_physical_plan_builder.go]: I think you either need to take just one condition as join condition or push all remaining equal condition down to child plan. Merge join will assume all eq condition keys are sorted accordingly. But here if only one key is pushed as required property, 2nd, 3rd and other remaining keys might be unsorted for the same 1st key group.

**Comment by hanfei1991** [comment]
@ilovesoup I will consider the order of multiple keys in the future.

File: plan/new_physical_plan_builder.go (hunk 1 above)
Comment (hanfei1991) [inline plan/new_physical_plan_builder.go]: @coocood @ilovesoup  I will add a TODO List here. This way is clearer and will be refactored in near future,

**Comment by coocood** [comment]
LGTM

**Comment by hanfei1991** [comment]
@ilovesoup PTAL

**Comment by ilovesoup** [comment]
For join key comment "I think you either need to take just one condition as join condition or push all remaining equal condition down to child plan. Merge join will assume all eq condition keys are sorted accordingly. But here if only one key is pushed as required property, 2nd, 3rd and other remaining keys might be unsorted for the same 1st key group."
is about correctness. Any thoughts? Even consider multi-key in the future, I guess need to do it correctly, right?

**Comment by ilovesoup** [comment]
LGTM

**Pull request merged**